[Improvement] PEES-836: Refactor Version CleanUp Task and reduce and fix slow queries

kingjia90 opened the pull request:
<!--

Before working on a contribution, you must determine on which branch you need to work:
- Bug fix: choose the latest maintenance branch `11.5`
- Feature/Improvement: choose `12.x` 

> All bug fixes merged into the latest maintenance branch are also merged to the current dev branch (`12.x`) on a regular basis.

## Please make sure your PR complies with all of the following points: 
- [ ] Read and accept our [contributing guidelines](/CONTRIBUTING.md) before you submit a PR.
- [ ] Features need to be proper documented in `doc/` 
- [ ] Bugfixes need a short guide how to reproduce them -> target branch is the oldest supported maintenance branch, e.g. `11.4` (see Readme.md for the list of supported versions)
- [ ] Meet all coding standards (see PhpStan actions) 

**Don't submit a PR if it doesn't comply, it'll be closed without a comment!**
-->  
  

## Changes in this pull request  
Partially Resolves https://github.com/pimcore/internal-improvements/issues/8, Resolves https://github.com/pimcore/service-operations/issues/282

## Additional info
Initially was a collection PR about reducing queries and fix slow queries, but happened to be a PR focused on Version Clean up after being brought up in an enterprise ticket, so some unrelated changes are still there.

kingjia90 committed:
use distinct instead of group by

kingjia90 committed:
id NOT IN (0) fix

kingjia90 committed:
refactor maintenanceGetOutdatedVersions to not be N + 1

kingjia90 commented on models/Version/Dao.php:
```diff
@@ -113,48 +113,55 @@ public function isBinaryHashInUse(?string $hash): bool
                         }
+                        $countsPerCid[$cid]++;
 
-                        if (count($versionIds) > 1000) {
```
i believe the stop and memory clear are not necessary anymore, the new query delegates all the resource hungry process to the DB (in a subquery) and is more efficient by avoiding a N+1 .
Need some tests with big data instance after creating many versions of the same element

kingjia90 replied:
Resolved: added a limit/offset way, should be even more lighter on memory

kingjia90 committed:
reduce frequency which last login date is updated, so the requests that have the exact timestamp or within 15 seoconds do not trigger it.

kingjia90 committed:
sonar fixes

kingjia90 committed:
sonar fixes

kingjia90 commented on lib/Tool/Authentication.php:
```diff
@@ -54,7 +54,11 @@ public static function authenticateSession(?Request $request = null): ?User
             if ($user instanceof \Pimcore\Security\User\User && self::isValidUser($user->getUser())) {
                 $pimcoreUser = $user->getUser();
-                $pimcoreUser->setLastLoginDate(); //set user current login date
+                // update last login date if last login timestamp was more than few seconds ago
```
this is to avoid to re-update the same value or updates in some sub-requests that happens few seconds apart, as we are not needing a real-time value

melissakittl commented on models/Version/Dao.php:
```diff
@@ -113,52 +113,63 @@ public function isBinaryHashInUse(?string $hash): bool
+                $versionIds = array_merge($versionIds, $tmpVersionIds);
+            } else {
+
+                $sql = '
+                    SELECT cid,id
+                    FROM (
+                        SELECT id, cid,
+                               ROW_NUMBER() OVER (PARTITION BY cid ORDER BY id DESC) AS rownumber
+                        FROM versions
+                        WHERE ctype = ? AND public = 0 ' . $ignoreIdsQueryPart . '
+                    ) sub
+                    WHERE rownumber > ?
+                ';
+
+                $countsPerCid = [];
+                $elementVersions = $this->db->fetchAssociative(
+                    $sql,
+                    [$elementType['elementType'], $elementType['steps'] + 1]
+                );
+
+                foreach ($elementVersions as $versionInfo) {
+                    $cid = $versionInfo['cid'];
+                    if (!isset($countsPerCid[$cid])) {
+                        $countsPerCid[$cid] = 0;
                     }
+                    $countsPerCid[$cid]++;
 
-                    if ($stop) {
-                        break;
-                    }
+                    $versionIds[] = $versionInfo['id'];
+                }
```
@kingjia90:
I think this is the wrong DB fetch method for multi-row result: `fetchAssociative` returns only a single row, but the code treats it as a list and iterates it. A better use would be `fetchAllAssociative` here. But with `fetchAllAssociative` and very large datasets and cids with thousands of versions, this implementation might become memory- and time-intensive, because it would load all matching rows into PHP memory.
So maybe back to the executeQuery and fetchAssociative combination, but with the new sql query and maybe a limit?

```suggestion
               $sql = '
                    SELECT cid,id
                    FROM (
                        SELECT id, cid,
                               ROW_NUMBER() OVER (PARTITION BY cid ORDER BY id DESC) AS rownumber
                        FROM versions
                        WHERE ctype = ? AND public = 0 ' . $ignoreIdsQueryPart . '
                    ) sub
                    WHERE rownumber > ?
                    LIMIT 1000
                ';
                $elementVersions = $this->db->executeQuery(
                    $sql,
                    [$elementType['elementType'], $elementType['steps'] + 1]
                );
                
                while ($versionInfo = $elementVersions->fetchAssociative()) {
                    $cid = $versionInfo['cid'];
                    if (!isset($countsPerCid[$cid])) {
                        $countsPerCid[$cid] = 0;
                    }
                    $countsPerCid[$cid]++;

                    $versionIds[] = $versionInfo['id'];
                }
```

kingjia90 replied:
Thank you for the feedback!
I've tried to apply some OFFSET/LIMIT on https://github.com/pimcore/pimcore/pull/18747/commits/4d3ec97d697358ed085d03803780734a490d35d4 

But overall, have it worked for you on real big data? i haven't tested it yet so far as i need to simulate a similar instance

jdreesen replied:
Why not `iterateAssociative` instead of `fetchAllAssociative` or `executeQuery` and then `fetchAssociative` with a `while` loop?

kingjia90 replied:
ahh yeah, thank you for the tip, it would be much better on the memory with iterateAssociative while being very slightly "slower", right? I've applied on https://github.com/pimcore/pimcore/pull/18747/commits/54170b3d404025ebdb66c37cc3e27f2dbcfa2d1e

melissakittl replied:
@kingjia90 
I tested it with a large dataset, and SQL queries are more or less faster than before. Below are the results for different <code>elementType</code> queries:

Element Type | Total Entries | Query Time | SQL Query
-- | -- | -- | --
documents | 78,716 entries | 533ms | `SELECT cid,id FROM (SELECT id, cid, ROW_NUMBER() OVER (PARTITION BY cid ORDER BY id DESC) AS rownumber FROM versions WHERE ctype = 'document' AND public = 0) sub WHERE rownumber > 5;`
assets | 1,443,556 entries | 4s | `SELECT cid,id FROM (SELECT id, cid, ROW_NUMBER() OVER (PARTITION BY cid ORDER BY id DESC) AS rownumber FROM versions WHERE ctype = 'asset' AND public = 0) sub WHERE rownumber > 5;`
objects | 11,774,799 entries | 50s | `SELECT cid,id FROM (SELECT id, cid, ROW_NUMBER() OVER (PARTITION BY cid ORDER BY id DESC) AS rownumber FROM versions WHERE ctype = 'object' AND public = 0) sub WHERE rownumber > 3;`


<ul>
<li>Total entries in the versions table: 13,297,071</li>
<li>The queries for <code>objects</code> take significantly longer (up to 50 seconds) due to the large number of entries. Not sure if this is an acceptable amount of time?</li>
<li>Additionally, the query did not include an <code>$ignoreIdsQueryPart</code> . Adding an ignore list would likely result in even longer query times.</li></ul>

kingjia90 replied:
thank you for the testing it and for the detailed performance report!

I am afraid that the time would be tied to the quantity (the element type is not that relevant here, objects are almost 8+ times of the assets and takes, let's say, 4-6s so for objects taking to up to 6sx8times=48s or so, it  seems just reasonable in this context). That was also my main worry, that testing with fake data would not fully reflect the real-life data.

The main focus of the PR was to reduce the queries and we kind of managed to improve the overall memory issues like usage and potential crashes while increasing the performance (former N+1 that lead to many slow round trips) . 
So it would be a nice starting point for further improvements, maybe LATERAL JOIN/CROSS JOINS.

BTW I've improved the maintenance task itself too
https://github.com/pimcore/pimcore/pull/18747/commits/1ee261ff3302856825b647252d04696560388587 while increasing a little more the expensiveness of the query itself but should be overall better.

Some ideas that pop up in my mind is maybe an integrity sanitation query of the "orphaned" versions (the one that was covered by https://github.com/pimcore/pimcore/blob/d3da104dd6d215ead2b2b8c1c394f6ff11627ba3/lib/Maintenance/Tasks/VersionsCleanupTask.php#L159)  to be run first, so that the amount of entries to check is reduced beforehand.

kingjia90 replied:
@melissakittl  if possible, could you please also check the amount of `getOrphanedVersions()` (introduced in https://github.com/pimcore/pimcore/pull/18747/commits/34cf346cf15d49d5caf901fd2d52cd77066cd8cf), just to have a rough idea/estimate of how much % of the total it could be.

melissakittl replied:
@kingjia90 
of course. Seems to be pretty fast, even for objects: 

Element Type  | Query Time | SQL Query | total entries | orphaned versions
-- | -- | -- | -- | --
Documents  | 450ms | `SELECT versions.id FROM versions LEFT JOIN documents AS element ON element.id = versions.cid WHERE element.id IS NULL AND versions.ctype = 'document';` | 78,716 entries | 5294 versions to delete
Assets  | 456ms | `SELECT versions.id FROM versions LEFT JOIN assets AS element ON element.id = versions.cid WHERE element.id IS NULL AND versions.ctype = 'asset';` | 1,443,556 entries | 456,455 versions to delete
Objects | 468ms | `SELECT versions.id FROM versions LEFT JOIN objects AS element ON element.id = versions.cid WHERE element.id IS NULL AND versions.ctype = 'object';` | 11,774,799 entries | 7,525,208 versions to delete

But if these results are correct, it seems I have a huge number of "orphaned" versions in the system that should probably be cleaned up… and the cleanup process hasn’t been working properly for quite some time

kingjia90 replied:
ahh thank you again for the report
So it's seems that deleting the orphaned first would help a lot overall, especially in this case. I was expecting that the percentage would be around 10 or so. not like 60%+. This leads to a question on why it's not deleted on element deletion but we will address it in a separate PR.

If my calculation are right, it would approximately take 12s (assets 1443556 rows took 4s, so objects (11774799−7525208)/1443556 = 2.94times x 4s = 12s) to perform the data object one after the deletion. 
In your case, the deletion of orphaned ones might take a while instead (it's a N+1 with 8 million entries) but i've introduced [this](https://github.com/pimcore/pimcore/pull/18747/commits/996b446a33dab29b0ffd92200d6845785699218f) that would help in avoid the N+1.

The former code was a while true loop, that's probably why it had a tendency to crash or timeout and ended up in this situation: it only checked the top occurences like 50k( just random number 50 times x 1000 limit) in each run, while building up behind more and more data that is not even touched because it stoppped or crashed before reaching that. Even by re-run, it would still re-process the top occurences again and again, of course minus the few deletions happened during each run.

kingjia90 committed:
Implement pagination for version retrieval query

kingjia90 committed:
phpstan fix

kingjia90 committed:
Refactor countsPerCid initialization and logging

kingjia90 committed:
Enhance logging for version count by element type

kingjia90 committed:
Update SQL query to use dynamic limit and offset

Refactor SQL query to include limit and offset parameters.

kingjia90 committed:
use iterativeAssociative approach

kingjia90 committed:
refactor to avoid loading elements and check N times scheduled task

kingjia90 commented on lib/Maintenance/Tasks/VersionsCleanupTask.php:
```diff
@@ -90,81 +90,21 @@ private function doVersionCleanup(): void
-            }
+        foreach ($versions as $index => $id) {
+            $version = Version::getById($id);
+            $version->delete();
```
this is kept to still trigger the PRE/POST DELETE events
https://github.com/pimcore/pimcore/blob/df14e025d3db615fc75b190e3c9f3cc1d68d7304/models/Version.php#L251

It would be even better if this foreach is replaced to implode `$versions` ids and delete by `WHERE IN`

kingjia90 commented on models/Version/Dao.php:
```diff
@@ -111,54 +111,75 @@ public function isBinaryHashInUse(?string $hash): bool
+                        $countsPerCid[$cid] = 0;
                     }
+                    $countsPerCid[$cid]++;
+                    $versionIds[] = $versionInfo['id'];
```
the CleanupTask is less verbose after this [commit](https://github.com/pimcore/pimcore/pull/18747/commits/1ee261ff3302856825b647252d04696560388587) but prevents from loading every element related to the version to be deleted and the scheduled tasks query https://github.com/pimcore/pimcore/blob/1ee261ff3302856825b647252d04696560388587/models/Version/Dao.php#L85


we could SELECT the information and still log the cases, eg.  not deleted because version is newer than actual or element didn't exist anymore etc..

kingjia90 replied:
for versions orphaned by a deleted element, just made a script to handle them before the heavy SQL
https://github.com/pimcore/pimcore/pull/18747/commits/34cf346cf15d49d5caf901fd2d52cd77066cd8cf
so only left is the one about newer version or being scheduled would be missing, but they do not look relevant for the task itself, rather for debugging and as safe guards (which are not necessary as long as the heavy query is correct)

kingjia90 commented on lib/Maintenance/Tasks/VersionsCleanupTask.php:
```diff
@@ -90,81 +90,21 @@ private function doVersionCleanup(): void
         }
 
-        $list = new Version\Listing();
-        $ignoredIds = $list->setLoadAutoSave(true)->setCondition(' autoSave = 1 ')->loadIdList();
```
moved as a condition in the `maintenanceGetOutdatedVersions` query , no need to instantiate a listing and pass the `$ignoredIds` anymore

kingjia90 committed:
stan fix

kingjia90 committed:
delete orphaned versions first to make it leaner for maintenanceGetOutdatedVersions

kingjia90 committed:
array of ids

kingjia90 committed:
sonar

kingjia90 commented on lib/Maintenance/Tasks/VersionsCleanupTask.php:
```diff
@@ -90,81 +90,32 @@ private function doVersionCleanup(): void
+            $this->logger->debug(
+                'delete version (' . $versionId . ") because the corresponding element doesn't exist anymore"
+            );
+            $version->delete();
```
in this case, the element which the version belong to doesn't exists anymore, i believe the PRE and POST delete event are not useful anymore unlike https://github.com/pimcore/pimcore/pull/18747/files#r2517696388 , should be a WHERE IN of the IDs with some slices of the `$orphanVersions` array.
Can't think of any particular use-case 🤔 maybe making it configurable to trigger these events or not would be the best of two Worlds.

kingjia90 committed:
configuration to disable PRE/POST delete events on Version, use WHERE IN

kingjia90 committed:
refactor

kingjia90 committed:
cs

kingjia90 committed:
Merge remote-tracking branch 'origin/12.x' into improve-sql

kingjia90 committed:
fix stan

kingjia90 committed:
Optimize last login update logic

Add check to update last login date only if more than 15 seconds have passed since last login.

kingjia90 committed:
fix config typo and inglobate doAutoSaveVersionCleanup into the new script
add docs

kingjia90 committed:
refactor orphaned to delete outdated autosave too

kingjia90 committed:
Merge branch '12.x' into improve-sql

kingjia90 commented on lib/Maintenance/Tasks/VersionsCleanupTask.php:
```diff
@@ -35,26 +35,8 @@ public function __construct(private LoggerInterface $logger, private SystemSetti
-        $this->doAutoSaveVersionCleanup();
     }
 
-    private function doAutoSaveVersionCleanup(): void
```
this N+1 could just be reduced by adding this condition to orphaned query https://github.com/pimcore/pimcore/pull/18747/files#diff-301ad5f95cdf128e949b26dc28992a2eed0bf8b0eb30012bda13435ab79c3bd8R196

Less readable and SRP but a good compromise for this case as it was formerly sensible to crashes and slowdown and it's for an internal clean up purpose

kingjia90 committed:
cs

kingjia90 committed:
Merge remote-tracking branch 'origin/12.x' into improve-sql

kingjia90 commented on models/Version/Dao.php:
```diff
@@ -107,60 +108,125 @@ public function isBinaryHashInUse(?string $hash): bool
+        $results = [];
+
+        $autoSaveDateCleanup = \Carbon\Carbon::now();
+        $autoSaveDateCleanup->subHours(72);
```
this should be configurable too... but out of the scope of this PR

kingjia90 committed:
Update Authentication.php
